File: dodam-application/dodam-rest-api/src/main/java/b1nd/dodam/restapi/notice/application/NoticeUseCase.java
package b1nd.dodam.restapi.notice.application;

import b1nd.dodam.domain.rds.division.entity.Division;
import b1nd.dodam.domain.rds.division.service.DivisionMemberService;
import b1nd.dodam.domain.rds.division.service.DivisionService;
import b1nd.dodam.domain.rds.member.entity.Member;
import b1nd.dodam.domain.rds.notice.entity.Notice;
import b1nd.dodam.domain.rds.notice.entity.NoticeDivision;
import b1nd.dodam.domain.rds.notice.entity.NoticeFile;
import b1nd.dodam.domain.rds.notice.enumration.NoticeStatus;
import b1nd.dodam.domain.rds.notice.service.NoticeService;
import b1nd.dodam.restapi.auth.infrastructure.security.support.MemberAuthenticationHolder;
import b1nd.dodam.restapi.notice.application.data.req.GenerateNoticeReq;
import b1nd.dodam.restapi.notice.application.data.req.ModifyNoticeReq;
import b1nd.dodam.restapi.notice.application.data.res.NoticeRes;
import b1nd.dodam.restapi.support.data.Response;
import b1nd.dodam.restapi.support.data.ResponseData;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
@Transactional(rollbackFor = Exception.class)
public class NoticeUseCase {

    private final NoticeService noticeService;
    private final DivisionService divisionService;
    private final DivisionMemberService divisionMemberService;
    private final MemberAuthenticationHolder memberAuthenticationHolder;

    public ResponseData<Long> register(GenerateNoticeReq generateNoticeReq) {
        Member member = memberAuthenticationHolder.current();
        Notice notice = noticeService.save(generateNoticeReq.toEntity(member));
        List<Division> divisions = divisionService.getAllByIds(generateNoticeReq.divisions());
        saveFilesAndDivisions(generateNoticeReq.toNoticeFiles(notice), generateNoticeReq.toNoticeDivisions(notice, divisions));
        return ResponseData.of(HttpStatus.OK, "공지 생성 성공", notice.getId());
    }

    private void saveFilesAndDivisions(List<NoticeFile> noticeFiles, List<NoticeDivision> noticeDivisions){
        noticeService.saveAllNoticeFiles(noticeFiles);
        noticeService.saveAll(noticeDivisions);
    }

    public Response modify(Long id, ModifyNoticeReq modifyNoticeReq){
        Member member = memberAuthenticationHolder.current();
        noticeService.updateNotice(id, member, modifyNoticeReq.title(), modifyNoticeReq.content());
        return Response.ok("공지 수정 성공");
    }

    @Transactional(readOnly = true)
    public ResponseData<List<NoticeRes>> getNotices(String keyword, Long lastId, int limit) {
        Member member = memberAuthenticationHolder.current();
        List<Long> divisionIds = divisionMemberService.getIdsByMember(member);
        List<Notice> notices = noticeService.getAllByStatus(keyword, divisionIds, lastId, limit);
        List<NoticeRes> noticeResList = convertToNoticeRes(notices);
        return ResponseData.of(HttpStatus.OK, "전체 공지 불러오기 성공", noticeResList);
    }

    @Transactional(readOnly = true)
    public ResponseData<List<NoticeRes>> getNoticesByDivision(Long divisionId, Long lastId, int limit) {
        Member member = memberAuthenticationHolder.current();
        List<Notice> notices = noticeService.getAllByDivision(member.getId(), divisionId, lastId, limit);
        List<NoticeRes> noticeResList = convertToNoticeRes(notices);
        return ResponseData.of(HttpStatus.OK, "카테고리별 공지 불러오기 성공", noticeResList);
    }

    public Response deleteNotice(Long id){
        Member member = memberAuthenticationHolder.current();
        noticeService.changeStatus(id, member, NoticeStatus.DELETED);
         return Response.ok("공지 삭제 성공");
    }

    private List<NoticeRes> convertToNoticeRes(List<Notice> notices) {
        Map<Long, List<NoticeFile>> noticeFileMap = noticeService.getNoticeFileMap(notices);
        return notices.stream()
                .map(notice -> NoticeRes.of(notice, noticeFileMap.getOrDefault(notice.getId(), List.of())))
                .collect(Collectors.toList());
    }

}

File: services/service-notice/src/main/kotlin/com/b1nd/dodamdodam/notice/application/notice/NoticeUseCase.kt
package com.b1nd.dodamdodam.notice.application.notice

import com.b1nd.dodamdodam.core.common.data.Response
import com.b1nd.dodamdodam.core.security.util.getCurrentUserId
import com.b1nd.dodamdodam.notice.application.notice.data.request.GenerateNoticeRequest
import com.b1nd.dodamdodam.notice.application.notice.data.request.ModifyNoticeRequest
import com.b1nd.dodamdodam.notice.application.notice.data.response.MemberInfoResponse
import com.b1nd.dodamdodam.notice.application.notice.data.response.NoticeResponse
import com.b1nd.dodamdodam.notice.application.notice.data.response.NoticeResponseMapper
import com.b1nd.dodamdodam.notice.domain.notice.entity.NoticeEntity
import com.b1nd.dodamdodam.notice.domain.notice.service.NoticeService
import com.b1nd.dodamdodam.notice.infrastructure.user.client.UserClient
import kotlinx.coroutines.runBlocking
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
@Transactional(rollbackFor = [Exception::class])
class NoticeUseCase(
    private val noticeService: NoticeService,
    private val userClient: UserClient,
    private val noticeResponseMapper: NoticeResponseMapper
) {
    fun register(request: GenerateNoticeRequest): Response<Long> {
        val userId = getCurrentUserId()
        val notice = noticeService.create(request.toEntity(userId))
        val files = request.toNoticeFiles(notice)
        if (files.isNotEmpty()) {
            noticeService.createFiles(files)
        }
        return Response.created("공지 생성 성공", notice.id)
    }

    @Transactional(readOnly = true)
    fun getNotices(keyword: String?, lastId: Long?, limit: Int): Response<List<NoticeResponse>> {
        val notices = noticeService.getAll(keyword, lastId, limit)
        val noticeResponses = convertToNoticeResponses(notices)
        return Response.ok("공지 목록 조회 성공", noticeResponses)
    }

    fun modify(id: Long, request: ModifyNoticeRequest): Response<Unit> {
        val userId = getCurrentUserId()
        noticeService.updateNotice(id, userId, request.title, request.content)
        return Response.ok("공지 수정 성공")
    }

    fun delete(id: Long): Response<Unit> {
        val userId = getCurrentUserId()
        noticeService.deleteNotice(id, userId)
        return Response.ok("공지 삭제 성공")
    }

    private fun convertToNoticeResponses(notices: List<NoticeEntity>): List<NoticeResponse> {
        val noticeFileMap = noticeService.getNoticeFileMap(notices)
        val userIds = notices.map { it.userId }.distinct()
        val memberInfoMap = fetchMemberInfos(userIds)

        return notices.map { notice ->
            noticeResponseMapper.toNoticeResponse(
                notice = notice,
                files = noticeFileMap.getOrDefault(notice.id!!, emptyList()),
                memberInfo = memberInfoMap[notice.userId]
            )
        }
    }

    private fun fetchMemberInfos(userIds: List<String>): Map<String, MemberInfoResponse> {
        if (userIds.isEmpty()) {
            return emptyMap()
        }
        return runBlocking {
            userClient.getUserInfosByUserIds(userIds)
        }
    }
}
